SLING-6516 Annotation Based ValidationModelProvider evaluating additional annotations on a Sling Model bound to a specific resource type and registering ValidationModels

karolisma opened the pull request:
Annotation Based ValidationModelProvider evaluating additional annotations on a Sling Model bound to a specific resource type and registering ValidationModels

karolisnc committed:
SLING-6516 Added ValidationModel registration on bundle installation.
Added ChildrenValidator, FieldValidator and NameRegex annotations.
Refactored ResourcePropertyBuilder and ChildResourceBuilder to work with annotations.
Added Sling Models API dependency.

karolisnc committed:
SLING-6516 Adding validation for child resources based on injected Sling Models.

karolisnc committed:
MAMASSETS-180 Refactoring. JUnit tests for ValidationModelImplementation.

karolisnc committed:
MAMASSETS-180 Fixing multiple and optional issues. JavaDocs. JUnit tests.

karolisnc committed:
SLING-6516 JavaDocs

karolisnc committed:
SLING-6516 Renaming class name to Validation Model Register.

karolisnc committed:
SLING-6516 Removing unused method. Minor changes in javadocs and method names.

kwin commented on bundles/extensions/validation/api/src/main/java/org/apache/sling/validation/annotations/Validate.java:
```diff
@@ -0,0 +1,63 @@
+     *
+     * @return validator properties
+     */
+    String[] properties() default {};
```
Since properties is a term which is used a lot, I would prefer naming it arguments. Also the javadoc can maybe something like "Parametrisation being passed to the validator given by the id. Each string entry must have the format `<key>=<value>`"

kwin commented on bundles/extensions/validation/api/src/main/java/org/apache/sling/validation/annotations/Validate.java:
```diff
@@ -0,0 +1,63 @@
+public @interface Validate {
+
+    /**
+     * Validator id, with which it is registered.
```
Rather "Validate the annotated element with a Validator having the given id".

kwin commented on bundles/extensions/validation/api/src/main/java/org/apache/sling/validation/annotations/Validate.java:
```diff
@@ -0,0 +1,63 @@
+     *
+     * @return the string
+     */
+    String validatorId() default "";
```
I am wondering why this can be empty. IMHO this annotation only makes sense in case the id is set as well, except for the case with the regex.

kwin commented on bundles/extensions/validation/api/src/main/java/org/apache/sling/validation/annotations/Validate.java:
```diff
@@ -0,0 +1,63 @@
+     *
+     * @return regex
+     */
+    String regex() default "";
```
Please clarify the javadoc here. Also since this can be used independent of an actual validator I would propose to rather come up with a dedicated annotation or not supporting this in the first place for Sling models (don't have a good use case in mind).

kwin commented on bundles/extensions/validation/api/src/main/java/org/apache/sling/validation/annotations/ValidationPaths.java:
```diff
@@ -0,0 +1,40 @@
+ */
+@Target(TYPE)
+@Retention(RUNTIME)
+public @interface ValidationPaths {
```
I think ApplicableValidationPaths is a clearer name.

kwin commented on bundles/extensions/validation/core/pom.xml:
```diff
@@ -171,6 +171,12 @@
+        <dependency>
+            <groupId>org.apache.sling</groupId>
+            <artifactId>org.apache.sling.models.api</artifactId>
+            <version>1.3.3-SNAPSHOT</version>
```
why does it depend on the newest snapshot? Please reference here the lowest models API you can work with.

kwin commented on bundles/extensions/validation/core/src/main/java/org/apache/sling/validation/impl/annotationmodel/AnnotationValidationModelProviderImpl.java:
```diff
@@ -0,0 +1,72 @@
+ * AnnotationValidationModelProviderImpl provides Validation Models based on Annotations.
+ * It is responsible for creation and registering of ValidationPackageBundleListener.
+ */
+@Component(immediate = true, service = ValidationModelProvider.class)
```
No need for immediate=true, it will be automatically activated once referenced.

kwin commented on bundles/extensions/validation/core/src/main/java/org/apache/sling/validation/impl/annotationmodel/ValidationModelRegister.java:
```diff
@@ -0,0 +1,113 @@
+import org.osgi.framework.Bundle;
+
+/** The Validation model register keeps all registered validation models. */
+final class ValidationModelRegister {
```
rather ValidationModelRegistry

kwin commented on bundles/extensions/validation/core/src/main/java/org/apache/sling/validation/impl/annotationmodel/AnnotationValidationModelProviderImpl.java:
```diff
@@ -0,0 +1,72 @@
+     */
+    @Activate
+    protected void activate(final ComponentContext ctx) {
+        this.listener = new ValidationPackageBundleListener(ctx.getBundleContext(), this.validationModelRegister);
```
How do you initially fill the registry with bundles which have been active before this bundle?

kwin commented on bundles/extensions/validation/core/src/main/java/org/apache/sling/validation/impl/annotationmodel/ValidationPackageBundleListener.java:
```diff
@@ -0,0 +1,192 @@
+
+/** The Validation package bundle listener. It analyzes all the bundles upon their activation, finds Sling Models marked with
+ * ValidationStrategy.REQUIRED or ValidationStrategy.OPTIONAL and registers/unregisters validation models. */
+public class ValidationPackageBundleListener implements BundleTrackerCustomizer {
```
I am wondering whether we can reuse some of the logic already contained in Sling Models Bundle listener? What about the former registers an event listener which is triggered by event being emitted from sling Models whenever a new model class is added/removed.

karolisnc committed:
SLING-6516 Removed regex from @Validate annotation.
Renamed @ValidationPaths to @ApplicableValidationPaths
Renamed ValidationModelRegister to ValidationModelRegistry
Improved JavaDocs.

karolisnc committed:
SLING-6516 Lowest possible Sling Models API version dependency.

karolisnc committed:
Merge branch 'trunk' of https://github.com/apache/sling into SLING-6516

* 'trunk' of https://github.com/apache/sling: (891 commits)
  SLING-7020 - DefaultDistributionComponentProvider is causing an NPE when agents are listed
  typo
  SLING-7009 NPE in ResourceChangeListenerInfo when resource.change.types is not String or String array
  [maven-release-plugin] prepare for next development iteration
  [maven-release-plugin] prepare release org.apache.sling.resourceresolver-1.5.30
  SLING-5755 - Complete Distribution documentation
  SLING-7019: Improve the MapEntries class by making sure we only use the resourceresolver while holding a lock and have not been disposed.
  SLING-7018: Fix a bug that removed to many aliases in certain cases when a resource got removed.
  SLING-5965 : added metrics for number of, oldest and timers for scheduler jobs, ability to group by thread pool or custom configurable filter, plus used all of this to build a HealthCheck out of it which complains if there is a scheduler job running for more than 60 seconds
  Add Apache Sling Service User Mapper 1.3.4, Resource Resolver 1.5.28, JCR Base 3.0.4, and JCR Resource 3.0.4
  SLING-7017 - Make it possible to mount packages at different paths
  Merge branch 'SLING-7015'
  Fix failing test, with Oak we don't have workspace info anymore
  update to latest fsresource versions
  update to latest fsresource version
  Set dependency to snapshot while vote is going on
  Set dependency to snapshot while vote is going on
  [maven-release-plugin] prepare for next development iteration
  [maven-release-plugin] prepare release org.apache.sling.jcr.resource-3.0.4
  [maven-release-plugin] prepare for next development iteration
  ...

karolisnc committed:
SLING-6516 Reverted javadocs arguments -> properties.

karolisnc committed:
Merge branch 'trunk' of https://github.com/apache/sling into SLING-6516

* 'trunk' of https://github.com/apache/sling: (32 commits)
  SLING-7027: Fix eta calculation for jobs and make sure we use the new eta on an updated.
  [maven-release-plugin] prepare for next development iteration
  [maven-release-plugin] prepare release org.apache.sling.security-1.1.6
  [maven-release-plugin] prepare for next development iteration
  [maven-release-plugin] prepare release sling-31
  SLING-7030 : ReferrerFilter broken as legacy SCR Annotations are not processed. Apply modified patch from  Dominique Jäggi
  SLING-6678 reference bndlib in the same version as being referenced from the maven-bundle-plugin 3.3.0
  SLING-6746 remove the deprecated Felix scr annotations from plugin mgmt
  SLING-6190 only try to modify natures in case something new was really added
  SLING-6190 improve exposure of Status via the Eclipse default status handler
  SLING-7029 - Extension - adding method to stop a bundle
  Update to latest resourceresolver 1.5.30
  Update to Launchpad Base 5.6.6-2.6.20
  releng: updated dependencies after staging release
  [maven-release-plugin] prepare for next development iteration
  [maven-release-plugin] prepare release htl-maven-plugin-1.0.8
  [maven-release-plugin] prepare for next development iteration
  [maven-release-plugin] prepare release org.apache.sling.scripting.sightly-1.0.36
  releng: updated htl compiler dependency
  [maven-release-plugin] prepare for next development iteration
  ...

rombert commented:
With the upcoming migration to git this repository will be deleted. If the pull request contains data that should be re-submitted I suggest that you save the changes locally and resubmit them after the migration, which will likely happen on Monday the 9th of October.
